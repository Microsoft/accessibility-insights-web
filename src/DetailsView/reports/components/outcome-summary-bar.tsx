// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { kebabCase } from 'lodash';
import * as React from 'react';

import { NamedSFC } from '../../../common/react/named-sfc';
import {
    outcomeIconMap,
    outcomeIconMapInverted,
    OutcomeStats,
    OutcomeType,
    outcomeTypeSemantics,
    OutcomeUnits,
} from './outcome-type';

export type OutcomeSummaryBarProps = {
    outcomeStats: Partial<OutcomeStats>,
    allOutcomeTypes: OutcomeType[],
    units?: OutcomeUnits,
    inverted?: boolean
}

export const OutcomeSummaryBar = NamedSFC<OutcomeSummaryBarProps>('OutcomeSummaryBar', (props) => {
    return (
        <div className="outcome-summary-bar">
            {props.allOutcomeTypes
                .map(outcomeType => {
                    const { units, inverted } = props;
                    const text = outcomeTypeSemantics[outcomeType].pastTense;
                    const iconMap = inverted === true ? outcomeIconMapInverted : outcomeIconMap;
                    const outcomeIcon = iconMap[outcomeType];
                    const count = props.outcomeStats[outcomeType];
                    const suffix = units === 'percentage' ? '% ' : '';

                    return (
                        <div key={outcomeType} style={{ flexGrow: count }}>
                            <span className={kebabCase(outcomeType)}>
                                {outcomeIcon} {count + suffix} <span className="outcome-past-tense">{text}</span>
                            </span>
                        </div>
                    );
                })}
        </div>
    );
});
