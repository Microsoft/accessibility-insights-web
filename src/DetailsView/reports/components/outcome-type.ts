// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { countBy, values } from 'lodash';

import { ManualTestStatus, ManualTestStatusData } from '../../../common/types/manual-test-status';

export type OutcomeType = 'pass' | 'incomplete' | 'fail';

export const allOutcomeTypes: OutcomeType[] = ['pass', 'incomplete', 'fail'];

export function outcomeTypeFromTestStatus(testStatus: ManualTestStatus) {
    const statusMap = {
        [ManualTestStatus.PASS]: 'pass',
        [ManualTestStatus.FAIL]: 'fail',
        [ManualTestStatus.UNKNOWN]: 'incomplete',
    } as { [P in ManualTestStatus]: OutcomeType };

    return statusMap[testStatus];
}

export function outcomeTypeSemanticsFromTestStatus(testStatus: ManualTestStatus): OutcomeTypeSemantic {
    return outcomeTypeSemantics[outcomeTypeFromTestStatus(testStatus)];
}

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'passed' },
    incomplete: { pastTense: 'incomplete' },
    fail: { pastTense: 'failed' },
};

export type OutcomeStats = { [OT in OutcomeType]: number };

export function outcomeStatsFromManualTestStatus(testStepStatus: ManualTestStatusData): OutcomeStats {
    const outcomeTypeSet = values(testStepStatus).map(s => outcomeTypeFromTestStatus(s.stepFinalResult));
    const stats = countBy(outcomeTypeSet) as OutcomeStats;
    stats.pass = stats.pass || 0;
    stats.incomplete = stats.incomplete || 0;
    stats.fail = stats.fail || 0;
    return stats;
}
