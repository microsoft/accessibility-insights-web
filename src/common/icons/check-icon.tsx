// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedFC } from '../react/named-fc';

export const CheckIcon = NamedFC('CheckIcon', () => (
    <span className="check-container">
        <svg
            width="16"
            height="16"
            viewBox="0 0 16 16"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            role="img"
            aria-hidden="true"
        >
            <circle cx="8" cy="8" r="8" fill="#228722" />
            <path
                fillRule="evenodd"
                clipRule="evenodd"
                d="M6.05904 11.1417L6.0616 11.1442C6.10737 11.19 6.15668 11.23 6.20867 11.2643C6.57256 11.5046 7.06707 11.4646 7.38742 11.1442C7.38861 11.143 7.38982 11.1418 7.391 11.1406L11.9312 6.60041C12.2974 6.23427 12.2974 5.64071 11.9312 5.27457C11.5651 4.90848 10.9715 4.90848 10.6054 5.27457L6.72452 9.15545L5.60041 8.03134C5.2343 7.66524 4.64071 7.66524 4.27459 8.03134C3.90847 8.39747 3.90847 8.99104 4.27459 9.35718L6.05904 11.1417Z"
                fill="var(--neutral-0)"
            />
        </svg>
    </span>
));

export const CheckIconInverted = NamedFC('CheckIconInverted', () => (
    <span className="check-container">
        <svg
            width="16"
            height="16"
            viewBox="0 0 16 16"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            role="img"
            aria-hidden="true"
        >
            <path
                fillRule="evenodd"
                clipRule="evenodd"
                d="M8 16C12.4183 16 16 12.4183 16 8C16 3.58172 12.4183 0 8 0C3.58172 0 0 3.58172 0 8C0 12.4183 3.58172 16 8 16ZM6.0616 11.1442L6.05904 11.1417L4.27459 9.35718C3.90847 8.99104 3.90847 8.39747 4.27459 8.03134C4.64071 7.66524 5.2343 7.66524 5.60041 8.03134L6.72452 9.15545L10.6054 5.27457C10.9715 4.90848 11.5651 4.90848 11.9312 5.27457C12.2974 5.64071 12.2974 6.23427 11.9312 6.60041L7.391 11.1406L7.38742 11.1442C7.06707 11.4646 6.57256 11.5046 6.20867 11.2643C6.15668 11.23 6.10737 11.19 6.0616 11.1442Z"
                fill="white"
            />
        </svg>
    </span>
));
