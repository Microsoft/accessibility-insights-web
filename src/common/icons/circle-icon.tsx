// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedFC } from '../react/named-sfc';

export const CircleIcon = NamedFC('CircleIcon', () => <span className="check-container" />);
