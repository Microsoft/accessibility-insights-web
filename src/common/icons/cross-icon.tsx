// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../react/named-sfc';

export const CrossIcon = NamedSFC('CrossIcon', () => (
    <span className="check-container">
        <span className="cross-icon" />
    </span>
));
