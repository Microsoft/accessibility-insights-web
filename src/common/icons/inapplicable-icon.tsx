// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../react/named-sfc';

export const InapplicableIcon = NamedSFC('InapplicableIcon', () => (
    <span className="check-container">
        <span className="inapplicable-icon" />
    </span>
));
