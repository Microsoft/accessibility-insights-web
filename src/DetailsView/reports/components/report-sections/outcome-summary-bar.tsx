// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { kebabCase } from 'lodash';
import * as React from 'react';

import { CheckIcon } from '../../../../common/icons/check-icon';
import { CrossIcon } from '../../../../common/icons/cross-icon';
import { InapplicableIcon } from '../../../../common/icons/inapplicable-icon';
import { NamedSFC } from '../../../../common/react/named-sfc';
import { SectionProps } from './report-section-factory';

const iconMap = {
    pass: <CheckIcon />,
    inapplicable: <InapplicableIcon />,
    fail: <CrossIcon />,
};

export const outcomeText = {
    pass: 'Passed',
    inapplicable: 'Not applicable',
    fail: 'Failed',
};

export type InstanceOutcomeType = 'pass' | 'fail' | 'inapplicable';
const allOutcomeTypes: InstanceOutcomeType[] = ['fail', 'pass', 'inapplicable'];

export type OutcomeSummaryBarProps = Pick<SectionProps, 'scanResult'>;

export const OutcomeSummaryBar = NamedSFC<OutcomeSummaryBarProps>('OutcomeSummaryBar', ({ scanResult }) => {
    const countSummary: { [type in InstanceOutcomeType]: number } = {
        pass: scanResult.passes.length,
        fail: scanResult.violations.reduce((total, violation) => {
            return total + violation.nodes.length;
        }, 0),
        inapplicable: scanResult.inapplicable.length,
    };

    return (
        <div className="outcome-summary-bar">
            {allOutcomeTypes.map(outcomeType => {
                const text = outcomeText[outcomeType];
                const outcomeIcon = iconMap[outcomeType];
                const count = countSummary[outcomeType];

                return (
                    <span key={outcomeType} className={kebabCase(outcomeType)} style={{ flexGrow: count }}>
                        {outcomeIcon} {count} {text}
                    </span>
                );
            })}
        </div>
    );
});
