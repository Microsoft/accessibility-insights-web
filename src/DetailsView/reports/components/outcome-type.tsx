// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { InapplicableIcon } from '../../../common/icons/inapplicable-icon';
import { CheckIcon } from './../../../common/icons/check-icon';
import { CircleIcon } from './../../../common/icons/circle-icon';
import { CrossIcon } from './../../../common/icons/cross-icon';
import { InstanceOutcomeType } from './report-sections/outcome-summary-bar';
import { RequirementOutcomeType } from './requirement-outcome-type';

export type OutcomeType = RequirementOutcomeType | InstanceOutcomeType;

export interface OutcomeTypeSemantic {
    pastTense: string;
}

export const outcomeTypeSemantics: { [OT in OutcomeType]: OutcomeTypeSemantic } = {
    pass: { pastTense: 'Passed' },
    incomplete: { pastTense: 'Incomplete' },
    fail: { pastTense: 'Failed' },
    inapplicable: { pastTense: 'Not applicable' },
};

export const outcomeIconMap: { [OT in OutcomeType]: JSX.Element } = {
    pass: <CheckIcon />,
    incomplete: <CircleIcon />,
    fail: <CrossIcon />,
    inapplicable: <InapplicableIcon />,
};
