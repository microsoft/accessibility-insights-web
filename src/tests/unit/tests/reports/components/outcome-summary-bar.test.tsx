// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { render } from '@testing-library/react';
import { CheckIcon, CheckIconInverted } from 'common/icons/check-icon';
import { CircleIcon } from 'common/icons/circle-icon';
import { CrossIcon, CrossIconInverted } from 'common/icons/cross-icon';
import { InapplicableIconInverted } from 'common/icons/inapplicable-icon';
import * as React from 'react';
import { OutcomeSummaryBar, OutcomeSummaryBarProps } from 'reports/components/outcome-summary-bar';
import { OutcomeType } from 'reports/components/outcome-type';
import { mockReactComponents } from 'tests/unit/mock-helpers/mock-module-helpers';

jest.mock('common/icons/check-icon');
jest.mock('common/icons/circle-icon');
jest.mock('common/icons/cross-icon');
jest.mock('common/icons/inapplicable-icon');

describe('OutcomeSummaryBar', () => {
    mockReactComponents([
        CrossIcon,
        CheckIcon,
        CheckIconInverted,
        CircleIcon,
        CrossIconInverted,
        InapplicableIconInverted,
    ]);
    const allOutcomeTypes: OutcomeType[] = ['pass', 'fail', 'incomplete'];
    const outcomeStats = {
        pass: 42,
        incomplete: 7,
        fail: 13,
    };

    it('show by percentage', () => {
        const props: OutcomeSummaryBarProps = { outcomeStats, allOutcomeTypes, countSuffix: '%' };
        const renderResult = render(<OutcomeSummaryBar {...props} />);

        expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it('show by count', () => {
        const props: OutcomeSummaryBarProps = { outcomeStats, allOutcomeTypes };
        const renderResult = render(<OutcomeSummaryBar {...props} />);

        expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it('render inverted badges', () => {
        const props: OutcomeSummaryBarProps = {
            outcomeStats,
            allOutcomeTypes,
            iconStyleInverted: true,
        };
        const renderResult = render(<OutcomeSummaryBar {...props} />);

        expect(renderResult.asFragment()).toMatchSnapshot();
    });

    it('show with text label', () => {
        const props: OutcomeSummaryBarProps = { outcomeStats, allOutcomeTypes, textLabel: true };
        const renderResult = render(<OutcomeSummaryBar {...props} />);

        expect(renderResult.asFragment()).toMatchSnapshot();
    });
});
