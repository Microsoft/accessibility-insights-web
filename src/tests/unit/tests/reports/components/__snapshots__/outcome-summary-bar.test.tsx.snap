// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutcomeSummaryBar render inverted badges 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass"
    >
      <span
        aria-hidden="true"
      >
        <CheckIconInverted />
      </span>
      42
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIconInverted />
      </span>
      13
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
    </span>
  </div>
</div>
`;

exports[`OutcomeSummaryBar show by count 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass"
    >
      <span
        aria-hidden="true"
      >
        <CheckIcon />
      </span>
      42
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIcon />
      </span>
      13
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
    </span>
  </div>
</div>
`;

exports[`OutcomeSummaryBar show by percentage 1`] = `
<div
  className="outcome-summary-bar"
>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 42,
      }
    }
  >
    <span
      className="pass"
    >
      <span
        aria-hidden="true"
      >
        <CheckIcon />
      </span>
      42
      %
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 13,
      }
    }
  >
    <span
      className="fail"
    >
      <span
        aria-hidden="true"
      >
        <CrossIcon />
      </span>
      13
      %
    </span>
  </div>
  <div
    className="outcome-summary"
    style={
      Object {
        "flexGrow": 7,
      }
    }
  >
    <span
      className="incomplete"
    >
      <span
        aria-hidden="true"
      >
        <CircleIcon />
      </span>
      7
      %
    </span>
  </div>
</div>
`;
